File: VB/GrowthWare.Framework/Common/NaturalComparer.vb
Imports System.Globalization

Namespace Common
    Public Class NaturalComparer
        Implements IComparer(Of String)
        Implements IComparer

        Private mParser1 As StringParser
        Private mParser2 As StringParser
        Private mNaturalComparerOptions As NaturalComparerOptions
        Private mSortAscending As Boolean = True

        Private Enum TokenType
            [Nothing]
            Numerical
            [String]
        End Enum


        Private Class StringParser
            Private mTokenType As TokenType
            Private mStringValue As String
            Private mNumericalValue As Decimal
            Private mIdx As Integer
            Private mSource As String
            Private mLen As Integer
            Private mCurChar As Char
            Private mNaturalComparer As NaturalComparer

            Sub New(ByVal naturalComparer As NaturalComparer)
                mNaturalComparer = naturalComparer
            End Sub

            Public Sub Init(ByVal source As String)
                If source Is Nothing Then source = String.Empty
                mSource = source
                mLen = source.Length
                mIdx = -1
                mNumericalValue = 0
                NextChar()
                NextToken()
            End Sub

            Public ReadOnly Property TokenType() As TokenType
                Get
                    Return mTokenType
                End Get
            End Property

            Public ReadOnly Property NumericalValue() As Decimal
                Get
                    If mTokenType = NaturalComparer.TokenType.Numerical Then
                        Return mNumericalValue
                    Else
                        Throw New NaturalComparerException("Internal Error: NumericalValue called on a non numerical value.")
                    End If
                End Get
            End Property

            Public ReadOnly Property StringValue() As String
                Get
                    Return mStringValue
                End Get
            End Property

            Public Sub NextToken()
                Do
                    'CharUnicodeInfo.GetUnicodeCategory 
                    If mCurChar = Nothing Then
                        mTokenType = NaturalComparer.TokenType.Nothing
                        mStringValue = Nothing
                        Exit Sub
                    ElseIf Char.IsDigit(mCurChar) Then
                        ParseNumericalValue()
                        Exit Sub
                        'ElseIf Char.IsLetter(mCurChar) Then
                        '    ParseString()
                        '    Exit Sub
                    Else
                        'ignore this character and loop some more
                        'NextChar()
                        ParseString()
                        Exit Sub
                    End If
                Loop
            End Sub

            Private Sub NextChar()
                mIdx += 1
                If mIdx >= mLen Then
                    mCurChar = Nothing
                Else
                    mCurChar = mSource(mIdx)
                End If
            End Sub

            Private Sub ParseNumericalValue()
                Dim start As Integer = mIdx
                Dim NumberDecimalSeparator As Char = NumberFormatInfo.CurrentInfo.NumberDecimalSeparator(0)
                Dim NumberGroupSeparator As Char = NumberFormatInfo.CurrentInfo.NumberGroupSeparator(0)
                Do
                    NextChar()
                    If mCurChar = NumberDecimalSeparator Then
                        ' parse digits after the Decimal Separator
                        Do
                            NextChar()
                            If Not Char.IsDigit(mCurChar) AndAlso mCurChar <> NumberGroupSeparator Then Exit Do
                        Loop
                        Exit Do
                    Else
                        If Not Char.IsDigit(mCurChar) AndAlso mCurChar <> NumberGroupSeparator Then Exit Do
                    End If
                Loop
                mStringValue = mSource.Substring(start, mIdx - start)
                If Decimal.TryParse(mStringValue, mNumericalValue) Then
                    mTokenType = NaturalComparer.TokenType.Numerical
                Else
                    ' We probably have a too long value
                    mTokenType = NaturalComparer.TokenType.String
                End If
            End Sub

            Private Sub ParseString()
                Dim start As Integer = mIdx
                Dim roman As Boolean = (mNaturalComparer.mNaturalComparerOptions And NaturalComparerOptions.RomanNumbers) <> 0
                Dim romanValue As Integer
                Dim lastRoman As Integer = Integer.MaxValue
                Dim cptLastRoman As Integer
                Do
                    If roman Then
                        Dim thisRomanValue As Integer = RomanLetterValue(mCurChar)
                        If thisRomanValue > 0 Then
                            Dim handled As Boolean = False

                            If (thisRomanValue = 1 OrElse thisRomanValue = 10 OrElse thisRomanValue = 100) Then
                                NextChar()
                                Dim nextRomanValue As Integer = RomanLetterValue(mCurChar)
                                If nextRomanValue = thisRomanValue * 10 Or nextRomanValue = thisRomanValue * 5 Then
                                    handled = True
                                    If nextRomanValue <= lastRoman Then
                                        romanValue += nextRomanValue - thisRomanValue
                                        NextChar()
                                        lastRoman = thisRomanValue \ 10
                                        cptLastRoman = 0
                                    Else
                                        roman = False
                                    End If
                                End If
                            Else
                                NextChar()
                            End If
                            If Not handled Then
                                If thisRomanValue <= lastRoman Then
                                    romanValue += thisRomanValue
                                    If lastRoman = thisRomanValue Then
                                        cptLastRoman += 1
                                        Select Case thisRomanValue
                                            Case 1, 10, 100
                                                If cptLastRoman > 4 Then roman = False
                                            Case 5, 50, 500
                                                If cptLastRoman > 1 Then roman = False
                                        End Select
                                    Else
                                        lastRoman = thisRomanValue
                                        cptLastRoman = 1
                                    End If
                                Else
                                    roman = False
                                End If
                            End If
                        Else
                            roman = False
                        End If
                    Else
                        NextChar()
                    End If
                    If Not Char.IsLetter(mCurChar) Then Exit Do
                Loop
                mStringValue = mSource.Substring(start, mIdx - start)
                If roman Then
                    mNumericalValue = romanValue
                    mTokenType = NaturalComparer.TokenType.Numerical
                Else
                    mTokenType = NaturalComparer.TokenType.String
                End If
            End Sub

        End Class

        Sub New(ByVal NaturalComparerOptions As NaturalComparerOptions)
            mNaturalComparerOptions = NaturalComparerOptions
            mParser1 = New StringParser(Me)
            mParser2 = New StringParser(Me)
        End Sub

        Sub New(ByVal NaturalComparerOptions As NaturalComparerOptions, ByVal Direction As NaturalComparerDirection)
            If Direction = NaturalComparerDirection.Descending Then mSortAscending = False
            mNaturalComparerOptions = NaturalComparerOptions
            mParser1 = New StringParser(Me)
            mParser2 = New StringParser(Me)
        End Sub

        Sub New()
            MyClass.New(NaturalComparerOptions.Default)
        End Sub

        Public Function Compare(ByVal string1 As String, ByVal string2 As String) As Integer Implements System.Collections.Generic.IComparer(Of String).Compare
            mParser1.Init(string1)
            mParser2.Init(string2)
            Dim result As Integer
            Do
                If mParser1.TokenType = TokenType.Numerical And mParser2.TokenType = TokenType.Numerical Then
                    ' both string1 and string2 are numerical 
                    result = Decimal.Compare(mParser1.NumericalValue, mParser2.NumericalValue)
                Else
                    result = String.Compare(mParser1.StringValue, mParser2.StringValue)
                End If
                If result <> 0 Then
                    ' if the sort direction is decending the reverse the result
                    If Not mSortAscending Then
                        If result = -1 Then result = 1 Else result = -1
                    End If
                    Return result
                Else
                    mParser1.NextToken()
                    mParser2.NextToken()
                End If
            Loop Until mParser1.TokenType = TokenType.Nothing And mParser2.TokenType = TokenType.Nothing
            Return 0 'identical
        End Function

        Private Shared Function RomanLetterValue(ByVal c As Char) As Integer
            Select Case c
                Case "I"c
                    Return 1
                Case "V"c
                    Return 5
                Case "X"c
                    Return 10
                Case "L"c
                    Return 50
                Case "C"c
                    Return 100
                Case "D"c
                    Return 500
                Case "M"c
                    Return 1000
                Case Else
                    Return 0
            End Select
        End Function

        Public Function RomanValue(ByVal string1 As String) As Integer
            mParser1.Init(string1)

            If mParser1.TokenType = TokenType.Numerical Then
                Return CInt(mParser1.NumericalValue)
            Else
                Return 0
            End If
        End Function

        Public Function IComparer_Compare(ByVal x As Object, ByVal y As Object) As Integer Implements System.Collections.IComparer.Compare
            Return Compare(DirectCast(x, String), DirectCast(x, String))
        End Function
    End Class

    Public Enum NaturalComparerOptions
        None
        RomanNumbers
        'DecimalValues <- we could put this as an option
        'IgnoreSpaces  <- we could put this as an option
        'IgnorePunctuation <- we could put this as an option
        [Default] = None
    End Enum

    <System.Flags()> Public Enum NaturalComparerDirection
        None
        Ascending
        Descending
    End Enum

    <Serializable()>
    Public Class NaturalComparerException
        Inherits Exception

        Sub New(ByVal msg As String)
            MyBase.New(msg)
        End Sub
    End Class
End Namespace

File: CS/GrowthWare.Framework/Common/NaturalComparer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;

namespace GrowthWare.Framework.Common
{
    /// <summary>
    /// Class NaturalComparer
    /// </summary>
    public class NaturalComparer : IComparer<string>, IComparer
    {

        private StringParser mParser1;
        private StringParser mParser2;
        private NaturalComparerOptions mNaturalComparerOptions;
        private bool mSortAscending = true;

        private enum TokenType
        {
            Nothing,
            Numerical,
            String
        }

        private class StringParser
        {
            private TokenType mTokenType;
            private string mStringValue;
            private decimal mNumericalValue;
            private int mIdx;
            private string mSource;
            private int mLen;
            private char mCurChar;
            private NaturalComparer mNaturalComparer;

            /// <summary>
            /// Initializes a new instance of the <see cref="StringParser" /> class.
            /// </summary>
            /// <param name="naturalComparer">The natural comparer.</param>
            public StringParser(NaturalComparer naturalComparer)
            {
                mNaturalComparer = naturalComparer;
            }

            /// <summary>
            /// Inits the specified source.
            /// </summary>
            /// <param name="source">The source.</param>
            public void Init(string source)
            {
                if (source == null) source = string.Empty;
                mSource = source;
                mLen = source.Length;
                mIdx = -1;
                mNumericalValue = 0;
                NextChar();
                NextToken();
            }

            /// <summary>
            /// Gets the type of the token.
            /// </summary>
            /// <value>The type of the token.</value>
            public TokenType TokenType
            {
                get { return mTokenType; }
            }

            /// <summary>
            /// Gets the numerical value.
            /// </summary>
            /// <value>The numerical value.</value>
            /// <exception cref="GrowthWare.Framework.Common.NaturalComparerException"></exception>
            public decimal NumericalValue
            {
                get
                {
                    if (mTokenType == NaturalComparer.TokenType.Numerical)
                    {
                        return mNumericalValue;
                    }
                    else
                    {
                        throw new NaturalComparerException("Internal Error: NumericalValue called on a non numerical value.");
                    }
                }
            }

            /// <summary>
            /// Gets the string value.
            /// </summary>
            /// <value>The string value.</value>
            public string StringValue
            {
                get { return mStringValue; }
            }

            /// <summary>
            /// Nexts the token.
            /// </summary>
            public void NextToken()
            {
                if (mCurChar == 0)
                {
                    mTokenType = NaturalComparer.TokenType.Nothing;
                    mStringValue = null;
                    return;
                }
                else if (char.IsDigit(mCurChar))
                {
                    ParseNumericalValue();
                    return;
                }
                else
                {
                    //ignore this character and loop some more
                    //NextChar()
                    ParseString();
                    return;
                }
            }

            private void NextChar()
            {
                mIdx += 1;
                if (mIdx >= mLen)
                {
                    mCurChar = '\0';
                }
                else
                {
                    mCurChar = mSource[mIdx];
                }
            }

            private void ParseNumericalValue()
            {
                int start = mIdx;
                char NumberDecimalSeparator = NumberFormatInfo.CurrentInfo.NumberDecimalSeparator[0];
                char NumberGroupSeparator = NumberFormatInfo.CurrentInfo.NumberGroupSeparator[0];
                do
                {
                    NextChar();
                    if (mCurChar == NumberDecimalSeparator)
                    {
                        // parse digits after the Decimal Separator
                        do
                        {
                            NextChar();
                            if (!char.IsDigit(mCurChar) && mCurChar != NumberGroupSeparator) break; // TODO: might not be correct. Was : Exit Do

                        }
                        while (true);
                        break; // TODO: might not be correct. Was : Exit Do
                    }
                    else
                    {
                        if (!char.IsDigit(mCurChar) && mCurChar != NumberGroupSeparator) break; // TODO: might not be correct. Was : Exit Do

                    }
                }
                while (true);
                mStringValue = mSource.Substring(start, mIdx - start);
                if (decimal.TryParse(mStringValue, out mNumericalValue))
                {
                    mTokenType = NaturalComparer.TokenType.Numerical;
                }
                else
                {
                    // We probably have a too long value
                    mTokenType = NaturalComparer.TokenType.String;
                }
            }

            private void ParseString()
            {
                int start = mIdx;
                bool roman = (mNaturalComparer.mNaturalComparerOptions & NaturalComparerOptions.RomanNumbers) != 0;
                int romanValue = 0;
                int lastRoman = int.MaxValue;
                int cptLastRoman = 0;
                do
                {
                    if (roman)
                    {
                        int thisRomanValue = NaturalComparer.RomanLetterValue(mCurChar);
                        if (thisRomanValue > 0)
                        {
                            bool handled = false;

                            if ((thisRomanValue == 1 || thisRomanValue == 10 || thisRomanValue == 100))
                            {
                                NextChar();
                                int nextRomanValue = NaturalComparer.RomanLetterValue(mCurChar);
                                if (nextRomanValue == thisRomanValue * 10 | nextRomanValue == thisRomanValue * 5)
                                {
                                    handled = true;
                                    if (nextRomanValue <= lastRoman)
                                    {
                                        romanValue += nextRomanValue - thisRomanValue;
                                        NextChar();
                                        lastRoman = thisRomanValue / 10;
                                        cptLastRoman = 0;
                                    }
                                    else
                                    {
                                        roman = false;
                                    }
                                }
                            }
                            else
                            {
                                NextChar();
                            }
                            if (!handled)
                            {
                                if (thisRomanValue <= lastRoman)
                                {
                                    romanValue += thisRomanValue;
                                    if (lastRoman == thisRomanValue)
                                    {
                                        cptLastRoman += 1;
                                        switch (thisRomanValue)
                                        {
                                            case 1:
                                            case 10:
                                            case 100:
                                                if (cptLastRoman > 4) roman = false;
                                                break;
                                            case 5:
                                            case 50:
                                            case 500:
                                                if (cptLastRoman > 1) roman = false;
                                                break;
                                        }
                                    }
                                    else
                                    {
                                        lastRoman = thisRomanValue;
                                        cptLastRoman = 1;
                                    }
                                }
                                else
                                {
                                    roman = false;
                                }
                            }
                        }
                        else
                        {
                            roman = false;
                        }
                    }
                    else
                    {
                        NextChar();
                    }
                    if (!char.IsLetter(mCurChar)) break; // TODO: might not be correct. Was : Exit Do

                }
                while (true);
                mStringValue = mSource.Substring(start, mIdx - start);
                if (roman)
                {
                    mNumericalValue = romanValue;
                    mTokenType = NaturalComparer.TokenType.Numerical;
                }
                else
                {
                    mTokenType = NaturalComparer.TokenType.String;
                }
            }

        }

        /// <summary>
        /// Initializes a new instance of the <see cref="NaturalComparer" /> class.
        /// </summary>
        /// <param name="NaturalComparerOptions">The natural comparer options.</param>
        public NaturalComparer(NaturalComparerOptions NaturalComparerOptions)
        {
            mNaturalComparerOptions = NaturalComparerOptions;
            mParser1 = new StringParser(this);
            mParser2 = new StringParser(this);
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="NaturalComparer" /> class.
        /// </summary>
        /// <param name="NaturalComparerOptions">The natural comparer options.</param>
        /// <param name="Direction">The direction.</param>
        public NaturalComparer(NaturalComparerOptions NaturalComparerOptions, NaturalComparerDirection Direction)
        {
            if (Direction == NaturalComparerDirection.Descending) mSortAscending = false;
            mNaturalComparerOptions = NaturalComparerOptions;
            mParser1 = new StringParser(this);
            mParser2 = new StringParser(this);
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="NaturalComparer" /> class.
        /// </summary>
        public NaturalComparer()
            : this(NaturalComparerOptions.Default)
        {
        }

        /// <summary>
        /// Compares the specified string1.
        /// </summary>
        /// <param name="string1">The string1.</param>
        /// <param name="string2">The string2.</param>
        /// <returns>System.Int32.</returns>
        public int Compare(string string1, string string2)
        {
            mParser1.Init(string1);
            mParser2.Init(string2);
            int result = 0;
            do
            {
                if (mParser1.TokenType == TokenType.Numerical & mParser2.TokenType == TokenType.Numerical)
                {
                    // both string1 and string2 are numerical 
                    result = decimal.Compare(mParser1.NumericalValue, mParser2.NumericalValue);
                }
                else
                {
                    result = string.Compare(mParser1.StringValue, mParser2.StringValue);
                }
                if (result != 0)
                {
                    // if the sort direction is decending the reverse the result
                    if (!mSortAscending)
                    {
                        if (result == -1) result = 1; else result = -1;
                    }
                    return result;
                }
                else
                {
                    mParser1.NextToken();
                    mParser2.NextToken();
                }
            }
            while (!(mParser1.TokenType == TokenType.Nothing & mParser2.TokenType == TokenType.Nothing));
            return 0;
            //identical
        }

        private static int RomanLetterValue(char c)
        {
            switch (c)
            {
                case 'I':
                    return 1;
                case 'V':
                    return 5;
                case 'X':
                    return 10;
                case 'L':
                    return 50;
                case 'C':
                    return 100;
                case 'D':
                    return 500;
                case 'M':
                    return 1000;
                default:
                    return 0;
            }
        }

        /// <summary>
        /// Romans the value.
        /// </summary>
        /// <param name="string1">The string1.</param>
        /// <returns>System.Int32.</returns>
        public int RomanValue(string string1)
        {
            mParser1.Init(string1);

            if (mParser1.TokenType == TokenType.Numerical)
            {
                return (int)mParser1.NumericalValue;
            }
            else
            {
                return 0;
            }
        }

        /// <summary>
        /// Is the comparer_ compare.
        /// </summary>
        /// <param name="x">The x.</param>
        /// <param name="y">The y.</param>
        /// <returns>System.Int32.</returns>
        public int IComparer_Compare(object x, object y)
        {
            return Compare((string)x, (string)x);
        }

        int System.Collections.IComparer.Compare(object x, object y)
        {
            return IComparer_Compare(x, y);
        }
    }

    /// <summary>
    /// Enum NaturalComparerOptions
    /// </summary>
    public enum NaturalComparerOptions
    {
        /// <summary>
        /// None
        /// </summary>
        None,

        /// <summary>
        /// RomanNumbers
        /// </summary>
        RomanNumbers,

        //DecimalValues <- we could put this as an option
        //IgnoreSpaces  <- we could put this as an option
        //IgnorePunctuation <- we could put this as an option

        /// <summary>
        /// Default
        /// </summary>
        Default = None
    }

    /// <summary>
    /// Enum NaturalComparerDirection
    /// </summary>
    [System.Flags()]
    public enum NaturalComparerDirection
    {
        /// <summary>
        /// None
        /// </summary>
        None,

        /// <summary>
        /// Ascending
        /// </summary>
        Ascending,
        /// <summary>
        /// Descending
        /// </summary>
        Descending
    }

    /// <summary>
    /// Class NaturalComparerException
    /// </summary>
    [Serializable]
    public class NaturalComparerException : Exception
    {

        /// <summary>
        /// Initializes a new instance of the <see cref="NaturalComparerException" /> class.
        /// </summary>
        /// <param name="msg">The MSG.</param>
        public NaturalComparerException(string msg)
            : base(msg)
        {
        }
    }
}
